### ISSUE TITLE
[Feature Request]: Date picker component enhancement: combined date and time picker

### ISSUE THREAD

**Issue by KevinCamelo** [carbon-design-system/carbon#10186]
### Summary

We're seeking an approved proposal to enhance Carbon's `date picker` component to support combined date and time inputs within the same fields.

This enhancement to `date picker` would be useful for when users need to filter down data between a specific date and time. It would also minimize the amount of space needed to display `time picker` and `single with calendar` side by side, while also creating a mental connection to help our users to understand which date corresponds to which time. 

| With AM & PM | Without AM & PM | 
| ----------- | ----------- | 
| ![date picker with meridiem](https://user-images.githubusercontent.com/29249866/143887374-5ac4ce4d-1a9f-4446-b8fa-7d59e0de9b09.gif) | ![date picker without meridiem](https://user-images.githubusercontent.com/29249866/143887428-3074788c-0950-4370-bed1-a8ad6b1410fa.gif) | 

### Justification

While date picker (and it's variants) work for many use cases, we've noticed that they can become overbearing when shown side by side when used to ask the user for a date and time range. In addition, when `date picker` is currently implemented, a visual disconnect is created between the date and the relevant time. 

| Current Process |
| ----------- | 
| ![Current Implementation](https://user-images.githubusercontent.com/29249866/143884015-0c307bec-cfaf-4dbf-ad3b-cfb780537c17.png) |

| Process with Proposed Enhancement |
| ----------- | 
| ![Proposal](https://user-images.githubusercontent.com/29249866/143884142-c1274c21-136c-40ed-8908-54bc04fcd95c.png) | 

Since this would be an enhancement on top of `date picker's` current offering, we are reaching out for development assistance.

### Desired UX and success metrics

The combined date and time enhancement to `Date picker` will provide a visual indicator for users to choose a date and a corresponding time. It's functionality should mirror Carbon's current `date picker` and `time picker`, as it just has been combined into one smaller component.

**Success metrics:**

- Users should be able to identify the ability to select a date from a calendar and its corresponding time.
- Change the way a calendar is displayed based on their pre-defined IBM Cloud settings. (A user in Ireland would likely not use AM or PM, thus they should be shown `Date and Time Picker without Meridiem`.
- Warn users when their choice invalid.

### Required functionality

The required functionality should mirror what is currently seen on [the documentation for Date Picker.](https://www.carbondesignsystem.com/components/date-picker/usage) However, one added requirement would be changing the way `time picker` and `date picker` are displayed based on a users predefined settings on IBM Cloud.

A user in Ireland would likely not use AM or PM, thus they should be shown `Date and Time Picker without Meridiem`.

### Specific timeline issues / requests

Three users have asked us for this refined functionality in the past two months. While there is no defined deadline, it would be nice to have as an option. We'd love to work with Carbon to make this happen.

### Available extra resources

Design resources: our team can create usage and style templates, create additional guidance, and help bring this to life!

Development resources: A developer has agreed to work with Cloud PAL to contribute this as an enhancement to Carbon.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)

**Comment by sstrubberg** [comment]
Hey @KevinCamelo, just wanted to give you a heads up that we had a discussion around this one during our Backlog Cleaning meeting this morning.

Here are some notes:

As we were walking through your spec, @tw15egan took it upon himself to do a quick POC in [Codesandbox](https://codesandbox.io/s/distracted-dewdney-neydr1?file=/src/App.js). We think this gets you about 95% to where you're looking to go. Can you confirm this?

Also, as-is, DatePicker/TimePicker is currently built on top of a technology called [Flatpickr](https://flatpickr.js.org/). This comes with a unique set of challenges to maintain and enhance our DatePicker/TimePicker component. Having said that, we are currently having conversation around the [future of datepicker](https://ibm.ent.box.com/notes/997427358581) of which we will definitely want to consider this spec as part of that.

**Comment by tw15egan** [comment]
There are issues with the example CodeSandbox; most notably, the `DatePicker` assumes the second child will be `DatePickerInput`. If we are to support this, we'd need to change `DatePicker` with a `type="range"` to look for `DatePickerInput` elements precisely, and render other children as-is

**Comment by sstrubberg** [comment]
@KevinCamelo Absolutely! I'll connect with you on Slack and we'll figure out what supporting this effort looks like.

**Comment by darenhrelic** [comment]
Hi, is there a timeline or priority set for this?  It seems like a no-brainer to combine the date and time into a single component..

My group needs a component like this for our upcoming QA release (which is why i'm asking).

Thank you!

Daren

**Comment by sstrubberg** [comment]
hey @darenhrelic, unfortunately, we had to deprioritize efforts regarding the Date/Time Picker because of competing work-streams. Don't worry though, I haven't forgotten about the DatePicker. Hopefully, we'll get to re-prioritize it soon.

**Comment by JordanWSmith15** [comment]
@sstrubberg The Sustainability Apps BU has created a version of this in our storybook. We have many scenarios where it is required; we have legacy applications that have database objects combining date and time data into a single attribute. Validation is run as a whole rather than individually against the application's business rules. Because of this, using 2 separate Carbon inputs to manage date and time is not a good UX for error handling. When one field is left null, it causes the other fields to error out. 
We have had challenges upgrading this custom component from Carbon V10 to V11. Please consider prioritizing this enhancement in the future.
Relates to https://github.com/carbon-design-system/carbon/issues/17370
![Image](https://github.com/user-attachments/assets/2d6e39ba-d71c-4550-93c7-11371b976d49)

**Comment by tay1orjones** [comment]
This could potentially be looped in as part of https://github.com/carbon-design-system/carbon/discussions/13355 if/when that is eventually tackled.
